File: implementation/scipamato/scipamato-core-persistence-jooq/src/test/java/ch/difty/scipamato/core/persistence/paper/searchorder/IntegerSearchTermEvaluatorTest.java
package ch.difty.scipamato.core.persistence.paper.searchorder;

import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Mockito.when;

import org.jooq.Condition;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import ch.difty.scipamato.core.entity.search.IntegerSearchTerm;
import ch.difty.scipamato.core.entity.search.IntegerSearchTerm.MatchType;

@ExtendWith(MockitoExtension.class)
@SuppressWarnings("SameParameterValue")
public class IntegerSearchTermEvaluatorTest extends SearchTermEvaluatorTest<IntegerSearchTerm> {

    private final IntegerSearchTermEvaluator e = new IntegerSearchTermEvaluator();

    @Mock
    private IntegerSearchTerm stMock;

    @Override
    protected IntegerSearchTermEvaluator getEvaluator() {
        return e;
    }

    private void expectSearchTerm(MatchType type, int v) {
        expectSearchTerm(type, v, null);
    }

    private void expectSearchTerm(MatchType type, int v1, Integer v2) {
        when(stMock.getType()).thenReturn(type);
        when(stMock.getFieldName()).thenReturn("fieldX");
        when(stMock.getValue()).thenReturn(v1);
        if (v2 != null)
            when(stMock.getValue2()).thenReturn(v2);
    }

    @Test
    public void buildingCondition_withGreaterThanComparison() {
        expectSearchTerm(MatchType.GREATER_THAN, 10);
        Condition c = e.evaluate(stMock);
        assertThat(c.toString()).isEqualTo("field_x > 10");
    }

    @Test
    public void buildingCondition_withGreaterThanOrEqualComparison() {
        expectSearchTerm(MatchType.GREATER_OR_EQUAL, 10);
        Condition c = e.evaluate(stMock);
        assertThat(c.toString()).isEqualTo("field_x >= 10");
    }

    @Test
    public void buildingCondition_withExactValue() {
        expectSearchTerm(MatchType.EXACT, 10);
        Condition c = e.evaluate(stMock);
        assertThat(c.toString()).isEqualTo("field_x = 10");
    }

    @Test
    public void buildingCondition_withLessThanOrEqualComparison() {
        expectSearchTerm(MatchType.LESS_OR_EQUAL, 10);
        Condition c = e.evaluate(stMock);
        assertThat(c.toString()).isEqualTo("field_x <= 10");
    }

    @Test
    public void buildingCondition_withLessThanComparison() {
        expectSearchTerm(MatchType.LESS_THAN, 10);
        Condition c = e.evaluate(stMock);
        assertThat(c.toString()).isEqualTo("field_x < 10");
    }

    @Test
    public void buildingCondition_withRangeComparison() {
        expectSearchTerm(MatchType.RANGE, 10, 15);
        Condition c = e.evaluate(stMock);
        assertThat(c.toString()).isEqualTo("field_x between 10 and 15");
    }

    @Test
    public void buildingCondition_withFieldValueMissing() {
        int any = 0;
        expectSearchTerm(MatchType.MISSING, any);
        Condition c = e.evaluate(stMock);
        assertThat(c.toString()).isEqualTo("field_x is null");
    }

    @Test
    public void buildingCondition_withAnyFieldValuePresent() {
        int any = 0;
        expectSearchTerm(MatchType.PRESENT, any);
        Condition c = e.evaluate(stMock);
        assertThat(c.toString()).isEqualTo("field_x is not null");
    }

}

File: implementation/scipamato/core/persistence-jooq/src/intTest/kotlin/ch/difty/scipamato/core/persistence/paper/searchorder/IntegerSearchTermEvaluatorTest.kt
package ch.difty.scipamato.core.persistence.paper.searchorder

import ch.difty.scipamato.core.entity.search.IntegerSearchTerm
import ch.difty.scipamato.core.entity.search.IntegerSearchTerm.MatchType
import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

@Suppress("FunctionName")
internal class IntegerSearchTermEvaluatorTest {

    private val evaluator = IntegerSearchTermEvaluator()

    private val stMock = mock<IntegerSearchTerm>()

    private fun expectSearchTerm(type: MatchType, v: Int) {
        expectSearchTerm(type, v, null)
    }

    private fun expectSearchTerm(type: MatchType, v1: Int, v2: Int?) {
        whenever(stMock.type).thenReturn(type)
        whenever(stMock.fieldName).thenReturn("fieldX")
        whenever(stMock.value).thenReturn(v1)
        if (v2 != null)
            whenever(stMock.value2).thenReturn(v2)
    }

    @Test
    fun buildingCondition_withGreaterThanComparison() {
        expectSearchTerm(MatchType.GREATER_THAN, 10)
        val c = evaluator.evaluate(stMock)
        assertThat(c.toString()).isEqualTo("field_x > 10")
    }

    @Test
    fun buildingCondition_withGreaterThanOrEqualComparison() {
        expectSearchTerm(MatchType.GREATER_OR_EQUAL, 10)
        val c = evaluator.evaluate(stMock)
        assertThat(c.toString()).isEqualTo("field_x >= 10")
    }

    @Test
    fun buildingCondition_withExactValue() {
        expectSearchTerm(MatchType.EXACT, 10)
        val c = evaluator.evaluate(stMock)
        assertThat(c.toString()).isEqualTo("field_x = 10")
    }

    @Test
    fun buildingCondition_withLessThanOrEqualComparison() {
        expectSearchTerm(MatchType.LESS_OR_EQUAL, 10)
        val c = evaluator.evaluate(stMock)
        assertThat(c.toString()).isEqualTo("field_x <= 10")
    }

    @Test
    fun buildingCondition_withLessThanComparison() {
        expectSearchTerm(MatchType.LESS_THAN, 10)
        val c = evaluator.evaluate(stMock)
        assertThat(c.toString()).isEqualTo("field_x < 10")
    }

    @Test
    fun buildingCondition_withRangeComparison() {
        expectSearchTerm(MatchType.RANGE, 10, 15)
        val c = evaluator.evaluate(stMock)
        assertThat(c.toString()).isEqualTo("field_x between 10 and 15")
    }

    @Test
    fun buildingCondition_withFieldValueMissing() {
        val any = 0
        expectSearchTerm(MatchType.MISSING, any)
        val c = evaluator.evaluate(stMock)
        assertThat(c.toString()).isEqualTo("field_x is null")
    }

    @Test
    fun buildingCondition_withAnyFieldValuePresent() {
        val any = 0
        expectSearchTerm(MatchType.PRESENT, any)
        val c = evaluator.evaluate(stMock)
        assertThat(c.toString()).isEqualTo("field_x is not null")
    }

    @Test
    fun buildingCondition_withIncompleteSearchTerm() {
        whenever(stMock.type).thenReturn(MatchType.INCOMPLETE)
        val c = evaluator.evaluate(stMock)
        assertThat(c.toString()).isEqualTo("1 = 0")
    }

    @Test
    fun buildingCondition_withFieldId_prependsTable() {
        expectSearchTerm(MatchType.GREATER_THAN, 10)
        whenever(stMock.fieldName).thenReturn("id")
        val c = evaluator.evaluate(stMock)
        assertThat(c.toString()).isEqualTo("paper.id > 10")
    }

}
